File: src/main/cpp/subsystems/ClimberRotate.cpp
#include "subsystems/ClimberRotate.h"
#include "Constants.h"


ClimberRotate::ClimberRotate() {
    // Create motor controllers
    rotation = new rev::CANSparkMax(
        CONSTANTS::MOTORS::CAN::CLIMBER_ROTATION_ID,
        rev::CANSparkMax::MotorType::kBrushless
    );


    // Setup current limit, idle mode, and encoder factor
    rotation->SetSmartCurrentLimit(
            CONSTANTS::CLIMBER::ROTATION::HARD_CURRENT_LIMIT.to<int>());
    rotation->SetSecondaryCurrentLimit(
            CONSTANTS::CLIMBER::ROTATION::SOFT_CURRENT_LIMIT.to<double>());
    rotation->GetEncoder().SetPositionConversionFactor(
            CONSTANTS::CLIMBER::ROTATION::TURN_TO_DEGREES.to<double>());


    // Set up PID controller
    rotation->GetPIDController().SetP(
            CONSTANTS::CLIMBER::ROTATION::PID::P);
    rotation->GetPIDController().SetI(
            CONSTANTS::CLIMBER::ROTATION::PID::I);
    rotation->GetPIDController().SetD(
            CONSTANTS::CLIMBER::ROTATION::PID::D);
    rotation->GetPIDController().SetFF(
            CONSTANTS::CLIMBER::ROTATION::PID::FF);
    rotation->GetPIDController().SetIZone(
            CONSTANTS::CLIMBER::ROTATION::PID::I_ZONE);
    rotation->GetPIDController().SetOutputRange(
            CONSTANTS::CLIMBER::ROTATION::PID::MIN,
            CONSTANTS::CLIMBER::ROTATION::PID::MAX
    );
}

ClimberRotate::~ClimberRotate() {
    delete rotation;
}

units::degree_t ClimberRotate::GetAngle() {
    return angle;
}

void ClimberRotate::SetAngle(units::degree_t angle) {
    // Save the angle
    ClimberRotate::angle = angle;

    // Update the PID controller
    rotation->GetPIDController().SetReference(
        angle.to<double>(), rev::CANSparkMaxLowLevel::ControlType::kPosition);
}

File: src/main/java/frc/bionicpolars/subsystems/climber/ClimberRotate.java
package frc.bionicpolars.subsystems.climber;

import com.revrobotics.CANSparkMax;
import com.revrobotics.RelativeEncoder;
import com.revrobotics.SparkMaxPIDController;

import frc.bionicpolars.Constants;

public class ClimberRotate {
    public static double GEARBOX_RATIO = 100.0,
            PULLEY_RATIO = 122.0 / 65.0,
            TURN_TO_DEGREE = 360 / (GEARBOX_RATIO * PULLEY_RATIO),
            P = 0.07,
            I = 0.0,
            D = 0.0;
    private CANSparkMax motor = new CANSparkMax(
            Constants.CLIMBER_ROTATION_ID,
            CANSparkMax.MotorType.kBrushless);
    private RelativeEncoder encoder = motor.getEncoder();
    private SparkMaxPIDController pid = motor.getPIDController();
    private double angle = 0.0;


    public ClimberRotate() {
        encoder.setPositionConversionFactor(TURN_TO_DEGREE);
        pid.setP(P);
        pid.setI(I);
        pid.setD(D);
        pid.setOutputRange(-1.0, 1.0);
    }

    public double get() {
        return angle;
    }

    public void set(double angle) {
        this.angle = angle;
        pid.setReference(angle, CANSparkMax.ControlType.kPosition);
    }

    public double getActual() {
        return encoder.getPosition();
    }

    public void zero() {
        encoder.setPosition(0.0);
    }
}
